File: src/render/camera.ts
/**
 * The camera. Owns the world->screen transform, the letterbox, and the shot kick.
 *
 * THE SCALE RULE IS A FAIRNESS CONSTRAINT, NOT A LAYOUT ONE (tuning.ts §presentation,
 * DESIGN.md §8.7). iOS ignores manifest `orientation` and gives web apps no JS orientation lock,
 * so rotating the phone must not buy sight-line:
 *
 *     scale         = min(vw, vh) / VIEW_MINOR_UNITS          // 440
 *     visible major = min(max(vw, vh) / scale, VIEW_MAJOR_MAX_UNITS)   // 900, excess letterboxed
 *
 * Derived from the SHORTER axis, so the field of view across the narrow dimension is identical
 * in portrait and landscape. The longer axis is clipped at 900 u. Max half-diagonal on any
 * supported device is then 500.9 u against SPAWN_RADIUS 560 - which is how the simulation gets
 * away with knowing nothing about the viewport.
 *
 * The camera writes NOTHING to World. It reads the player's interpolated position and that is all.
 */

import { VIEW_MAJOR_MAX_UNITS, VIEW_MINOR_UNITS } from '../core/index.js';

/** Shot kick: 4 px opposite the barrel, ~90 ms ease-out (DESIGN.md §10.5). */
const KICK_PIXELS = 4;
const KICK_DECAY_SEC = 0.09;
/**
 * SHAKE FREQUENCIES, Hz, and they are deliberately not the same number.
 *
 * Two axes shaken at one frequency trace a LINE - the camera slides back and forth along a
 * diagonal, which reads as a drag rather than as a jolt. Two frequencies that are close but not
 * harmonically related draw a Lissajous figure instead, so the viewport moves in a way that has no
 * direction in it. 31 and 23 are not tuned numbers; they are coprime, well above the eye's ability
 * to follow, and comfortably below the frame rate so a shake cannot alias into a slow wobble.
 */
const SHAKE_HZ_X = 31;
const SHAKE_HZ_Y = 23;
/** Extra world units queried around the camera rect when culling, ~= the largest sprite radius. */
const CULL_MARGIN = 80;

export class Camera {
  /** Viewport in CSS px. Source of truth is `visualViewport`, not `window.inner*`. */
  viewW = 1;
  viewH = 1;
  /** CSS px per world unit. */
  scale = 1;

  /** Camera centre, world units. Follows the player exactly - no lag, no lookahead. */
  x = 0;
  y = 0;

  /** Half-extent of the DRAWN (post-letterbox) world rect, in world units. */
  halfW = VIEW_MINOR_UNITS / 2;
  halfH = VIEW_MINOR_UNITS / 2;

  /** Letterbox bar thickness in CSS px on each side. One of these is always 0. */
  barX = 0;
  barY = 0;

  private kickX = 0;
  private kickY = 0;
  /** Peak shake amplitude in CSS px, and the clock it is driven and faded by. */
  private shakePx = 0;
  private shakeLeft = 0;
  private shakeTotal = 0;
  private shakeAge = 0;

  /**
   * Recomputes scale and letterbox for a new viewport. Cheap and idempotent - safe to call from
   * a debounced resize handler on every burst iOS emits during toolbar collapse.
   */
  resize(w: number, h: number): void {
    this.viewW = w > 1 ? w : 1;
    this.viewH = h > 1 ? h : 1;

    const minor = Math.min(this.viewW, this.viewH);
    const major = Math.max(this.viewW, this.viewH);
    this.scale = minor / VIEW_MINOR_UNITS;

    const majorUnits = Math.min(major / this.scale, VIEW_MAJOR_MAX_UNITS);
    // Excess on the major axis, in CSS px, split evenly between the two bars.
    const excessPx = major - majorUnits * this.scale;

    if (this.viewW >= this.viewH) {
      this.halfW = majorUnits / 2;
      this.halfH = VIEW_MINOR_UNITS / 2;
      this.barX = excessPx / 2;
      this.barY = 0;
    } else {
      this.halfW = VIEW_MINOR_UNITS / 2;
      this.halfH = majorUnits / 2;
      this.barX = 0;
      this.barY = excessPx / 2;
    }
  }

  /** Snap to a position with no interpolation. Used on run start. */
  snapTo(x: number, y: number): void {
    this.x = x;
    this.y = y;
    this.kickX = 0;
    this.kickY = 0;
    this.shakePx = 0;
    this.shakeLeft = 0;
    this.shakeTotal = 0;
    this.shakeAge = 0;
  }

  follow(x: number, y: number): void {
    this.x = x;
    this.y = y;
  }

  /**
   * Adds a recoil impulse opposite a unit direction. Purely cosmetic: it moves the transform,
   * never the world, so it cannot perturb the simulation.
   */
  kick(dirX: number, dirY: number): void {
    this.kickX -= dirX * KICK_PIXELS;
    this.kickY -= dirY * KICK_PIXELS;
  }

  /**
   * SHAKES THE WHOLE VIEWPORT for `seconds`, starting at `pixels` and fading linearly to nothing.
   *
   * A different thing from `kick`, not a bigger one. A kick is DIRECTIONAL - it says a shell left
   * the barrel that way - and it is over in 90 ms. A shake has no direction and lasts long enough
   * to be felt rather than glimpsed: it says the machine itself was hit. Nothing in the game shook
   * before this, which is exactly why it is worth spending on the one event that has to land.
   *
   * The LOUDEST of the shake and whatever is already running wins, rather than the sum: two events
   * on the same frame should not be able to throw the camera twice as far as either was authored to.
   */
  shake(pixels: number, seconds: number): void {
    if (pixels <= this.shakePx && this.shakeLeft > 0) return;
    this.shakePx = pixels;
    this.shakeLeft = seconds;
    this.shakeTotal = seconds;
    this.shakeAge = 0;
  }

  /**
   * Exponential ease-out on the kick, linear fade on the shake. `dtSec` is REAL time - effects are
   * not tick-locked, which is what lets both of these keep running while the simulation is frozen.
   */
  update(dtSec: number): void {
    const k = Math.exp(-dtSec / KICK_DECAY_SEC);
    this.kickX *= k;
    this.kickY *= k;
    if (Math.abs(this.kickX) < 0.01) this.kickX = 0;
    if (Math.abs(this.kickY) < 0.01) this.kickY = 0;

    if (this.shakeLeft > 0) {
      this.shakeAge += dtSec;
      this.shakeLeft -= dtSec;
      if (this.shakeLeft <= 0) {
        this.shakeLeft = 0;
        this.shakePx = 0;
      }
    }
  }

  /**
   * Current shake offset. Derived from the clock rather than integrated, so it cannot drift and it
   * always ends at exactly zero - an integrated shake left the world a pixel or two off centre for
   * the rest of the run, which nothing on screen ever explains.
   */
  private get shakeX(): number {
    if (this.shakeLeft <= 0) return 0;
    return Math.sin(this.shakeAge * SHAKE_HZ_X) * this.shakePx * (this.shakeLeft / this.shakeTotal);
  }

  private get shakeY(): number {
    if (this.shakeLeft <= 0) return 0;
    return Math.cos(this.shakeAge * SHAKE_HZ_Y) * this.shakePx * (this.shakeLeft / this.shakeTotal);
  }

  /** Screen-space x (CSS px) of the world container's origin. */
  get originX(): number {
    return this.viewW * 0.5 - this.x * this.scale + this.kickX + this.shakeX;
  }

  get originY(): number {
    return this.viewH * 0.5 - this.y * this.scale + this.kickY + this.shakeY;
  }

  /**
   * Broad-phase cull. We already know the camera rect, so we test here rather than paying for
   * Pixi's `CullerPlugin`, which would re-derive it per container (DESIGN.md §10.5).
   */
  isVisible(x: number, y: number, radius: number): boolean {
    const m = radius + CULL_MARGIN;
    return (
      x > this.x - this.halfW - m &&
      x < this.x + this.halfW + m &&
      y > this.y - this.halfH - m &&
      y < this.y + this.halfH + m
    );
  }
}

File: cs/src/Scrapyard.Game/Camera.cs
using Microsoft.Xna.Framework;

namespace Scrapyard.Game;

/// <summary>
/// The world-to-screen transform and the letterbox. Port of <c>src/render/camera.ts</c>.
/// </summary>
/// <remarks>
/// <para>
/// THE SCALE RULE IS A FAIRNESS CONSTRAINT, NOT A LAYOUT ONE. The original ships to phones, where
/// iOS gives web apps no orientation lock - so rotating the device must not buy sight-line:
/// </para>
/// <code>
/// scale         = min(vw, vh) / VIEW_MINOR_UNITS              // 440
/// visible major = min(max(vw, vh) / scale, VIEW_MAJOR_MAX)     // 900, excess letterboxed
/// </code>
/// <para>
/// Derived from the SHORTER axis, so the field of view across the narrow dimension is identical in
/// portrait and landscape. That is what caps the half-diagonal at 500.9 u against a spawn radius of
/// 560 - which is how the simulation gets away with knowing nothing about the viewport. It holds on
/// a desktop window for the same reason: a player who drags the window wider must not see further
/// than one who did not.
/// </para>
/// <para>
/// THE CAMERA WRITES NOTHING TO THE WORLD. It reads the player's interpolated position and that is
/// all.
/// </para>
/// </remarks>
public sealed class Camera
{
    public const double ViewMinorUnits = 440;
    public const double ViewMajorMaxUnits = 900;

    public double X { get; private set; }
    public double Y { get; private set; }

    /// <summary>Pixels per world unit.</summary>
    public double Scale { get; private set; } = 1;

    public double HalfW { get; private set; } = ViewMinorUnits / 2;
    public double HalfH { get; private set; } = ViewMinorUnits / 2;

    /// <summary>The letterbox bars, in pixels, on whichever axis has the excess.</summary>
    public double BarX { get; private set; }

    public double BarY { get; private set; }

    private int _viewW = 1;
    private int _viewH = 1;

    /// <summary>Shot kick: pixels opposite the barrel, easing out over about ninety milliseconds.</summary>
    private const double KickPixels = 4;

    private const double KickDecaySec = 0.09;

    /// <summary>
    /// SHAKE FREQUENCIES, Hz, and they are deliberately not the same number.
    /// </summary>
    /// <remarks>
    /// Two axes shaken at one frequency trace a LINE - the camera slides back and forth along a
    /// diagonal, which reads as a drag rather than as a jolt. Two frequencies that are close but not
    /// harmonically related draw a Lissajous figure instead, so the viewport moves in a way that has
    /// no direction in it. 31 and 23 are coprime and well above the eye's ability to follow either.
    /// </remarks>
    private const double ShakeHzX = 31;

    private const double ShakeHzY = 23;

    private double _kickX;
    private double _kickY;
    private double _shakePx;
    private double _shakeLeft;
    private double _shakeTotal;
    private double _shakeAge;

    public void Resize(int w, int h)
    {
        _viewW = w > 1 ? w : 1;
        _viewH = h > 1 ? h : 1;
        double minor = System.Math.Min(_viewW, _viewH);
        double major = System.Math.Max(_viewW, _viewH);
        Scale = minor / ViewMinorUnits;
        double majorUnits = System.Math.Min(major / Scale, ViewMajorMaxUnits);
        double excessPx = major - majorUnits * Scale;

        if (_viewW >= _viewH)
        {
            HalfW = majorUnits / 2;
            HalfH = ViewMinorUnits / 2;
            BarX = excessPx / 2;
            BarY = 0;
        }
        else
        {
            HalfW = ViewMinorUnits / 2;
            HalfH = majorUnits / 2;
            BarX = 0;
            BarY = excessPx / 2;
        }
    }

    public void Follow(double x, double y)
    {
        X = x;
        Y = y;
    }

    /// <summary>Snaps, dropping any kick or shake in progress. Used when a run starts.</summary>
    public void SnapTo(double x, double y)
    {
        X = x;
        Y = y;
        _kickX = 0;
        _kickY = 0;
        _shakePx = 0;
        _shakeLeft = 0;
        _shakeTotal = 0;
        _shakeAge = 0;
    }

    /// <summary>A shot's recoil: the view is pushed OPPOSITE the barrel.</summary>
    public void Kick(double dirX, double dirY)
    {
        _kickX -= dirX * KickPixels;
        _kickY -= dirY * KickPixels;
    }

    /// <summary>
    /// A jolt. A bigger shake replaces a smaller one; a smaller one does not interrupt a bigger.
    /// </summary>
    /// <remarks>
    /// Otherwise a stream of small hits during a boss's slam would keep resetting the slam's own
    /// shake to something feeble, and the biggest event on screen would be the one you felt least.
    /// </remarks>
    public void Shake(double pixels, double seconds)
    {
        if (pixels <= _shakePx && _shakeLeft > 0) return;
        _shakePx = pixels;
        _shakeLeft = seconds;
        _shakeTotal = seconds;
        _shakeAge = 0;
    }

    /// <summary>Advances the kick and the shake by WALL-CLOCK seconds. Both are presentation.</summary>
    public void Update(double dt)
    {
        double k = System.Math.Exp(-dt / KickDecaySec);
        _kickX *= k;
        _kickY *= k;
        if (System.Math.Abs(_kickX) < 0.01) _kickX = 0;
        if (System.Math.Abs(_kickY) < 0.01) _kickY = 0;

        if (_shakeLeft > 0)
        {
            _shakeAge += dt;
            _shakeLeft -= dt;
            if (_shakeLeft <= 0)
            {
                _shakeLeft = 0;
                _shakePx = 0;
            }
        }
    }

    private double ShakeX => _shakeLeft <= 0
        ? 0
        : System.Math.Sin(_shakeAge * ShakeHzX) * _shakePx * (_shakeLeft / _shakeTotal);

    private double ShakeY => _shakeLeft <= 0
        ? 0
        : System.Math.Cos(_shakeAge * ShakeHzY) * _shakePx * (_shakeLeft / _shakeTotal);

    /// <summary>
    /// World units to screen pixels, kick and shake included.
    /// </summary>
    /// <remarks>
    /// BOTH ARE SCREEN-SPACE, added here rather than to <see cref="X"/> and <see cref="Y"/>. A kick
    /// that moved the camera's world position would move what is CULLED and what the letterbox
    /// clips, so a hard enough shot would pop scenery in and out at the edges. It is a wobble of the
    /// picture, not of the viewpoint.
    /// </remarks>
    public Vector2 ToScreen(double wx, double wy) => new(
        (float)(BarX + (wx - X + HalfW) * Scale + _kickX + ShakeX),
        (float)(BarY + (wy - Y + HalfH) * Scale + _kickY + ShakeY));

    /// <summary>
    /// The world-space rectangle currently on screen, padded by <paramref name="pad"/> units so a
    /// sprite whose CENTRE is just off screen but whose body is not still gets drawn.
    /// </summary>
    public (double X0, double Y0, double X1, double Y1) VisibleBounds(double pad) =>
        (X - HalfW - pad, Y - HalfH - pad, X + HalfW + pad, Y + HalfH + pad);
}
